### PR TITLE
feat: config to overwrite headers for csv reader

### LINKED ISSUE #807: CSV headers needs the ability to overwrite the headers
The current behaviour is to add a headers row before the main data, which works if the file is set up that way. 

But if the columns are fixed, we can reduce a potential mapping step and replace the would-be fields to a fixed list of headers. In particular this is to resolve the limitation of referencing fields with spaces and symbols which normally couldn't be referenced through the expression language syntax, such as `field.with.dot`, `Field Name With Spaces`, etc.

Potentially other odd edge cases as well, which this would resolve.

### PR DESCRIPTION
**Pull request by keevan**
Resolves #807

image
This is what it looks like.

Added a test to ensure it works as expected.

### CODE REVIEW

File: classes/local/step/reader_csv.php [hunk 1]
```diff
@@ -81,6 +83,12 @@ public function csv_contents_generator() {
                 return;
             }
 
+            // Used the configured headers in place of the first row if overwriteheaders is true.
```
Comment (Peterburnett) [inline classes/local/step/reader_csv.php:86]: Will this lose data if there arent headers configured in the file?

File: tests/tool_dataflows_reader_csv_test.php [hunk 2]
```diff
@@ -159,12 +162,60 @@ public function test_bad_input_stream() {
...
+        $reader->config = Yaml::dump([
+            'path' => $this->inputpath,
+            'headers' => implode(',', $headers),
+            'overwriteheaders' => '1',
+            'delimiter' => ',',
+        ]);
+
+        // Create test input file.
+        $data = [
+            ['apple', 'berry', 'cucumber'],
```
Comment (Peterburnett) [inline tests/tool_dataflows_reader_csv_test.php:182]: can we also just add a test for a file with no headers, but override headers set

Comment (Peterburnett) [reply tests/tool_dataflows_reader_csv_test.php:182]: ignore, CSV is untyped, this isnt easy to do

File: classes/local/step/reader_csv.php (hunk 1 above)
Comment (Peterburnett) [reply classes/local/step/reader_csv.php:86]: Probably, but this is fine. ignore comment

File: classes/local/step/reader_csv.php [hunk 3]
```diff
@@ -42,6 +42,7 @@ public static function form_define_fields(): array {
         return [
             'path' => ['type' => PARAM_TEXT, 'required' => true],
             'headers' => ['type' => PARAM_TEXT],
+            'overwriteheaders' => ['type' => PARAM_TEXT],
```
Comment (Peterburnett) [inline classes/local/step/reader_csv.php:45]: This is a checkbox, should it be PARAM_BOOL?

File: classes/local/step/reader_csv.php [hunk 4]
```diff
@@ -62,6 +63,7 @@ public function csv_contents_generator() {
         $maxlinelength = 1000;
         $config = $this->get_variables()->get('config');
         $strheaders = $config->headers;
+        $overwriteheaders = !empty($config->overwriteheaders);
```
Comment (Peterburnett) [inline classes/local/step/reader_csv.php:66]: If swapped to PARAM bool, can we just treat that as a bool? rather than empty()

Comment (keevan) [reply classes/local/step/reader_csv.php:66]: The field might not exist yet. This just supports both and every other scenario.

**Review by Peterburnett — APPROVED** [review]
